src/test/java/schools/SchoolTest.java

```diff
@@ -27,76 +25,110 @@ public void clearDataAfterTest(){
         school.addClass(clazzA);
+        // assertThat(school.howManyClasses()).isEqual(0);
 
         // when
+        // to nie jest najszczesliwsza logika
+        // lepiej by bylo dodac
+        // school.hasClass(clazzA);
         boolean result = school.getSchoolClasses().containsValue(clazzA);
 
         // then
+
+        // to sprawdzenie jest ok
         assertTrue(result);
         assertThat(school.getSchoolClasses()).hasSize(1);
     }
     @Test
-    public void shouldReturnTrueIfRemovingSchoolClassSuccessful(){
+    public void shouldRemoveExistingClass(){
         // given
         school.addClass(clazzA);
 
         // when
+        // o nie jest test tego co bylo opisane wyzej - to jest nadal element given i lepiej by bylo zrobic to przez
+        // hasClass
         assertThat(school.getSchoolClasses()).containsValue(clazzA);
 
         // then
-        assertTrue(school.removeClass(clazzA));
+        // to jest when i lepiej by wygladalo
+        var result = school.removeClass(clazzA);
+
+        // then wlasciwe - do poprawy asercja na assertThat()
+        assertTrue(result);
         assertThat(school.getSchoolClasses()).hasSize(0);
     }
     @Test
-    public void shouldReturnFalseIfRemovingSchoolClassDoNotExist(){
+    public void shouldNotRemoveClassIfNotInSchoolAndLetContinue(){
         // given
-        school.addClass(clazzA);
-        assertThat(school.getSchoolClasses()).containsValue(clazzA);
-        school.removeClass(clazzA);
+        // tu bym nic nie robila, bo wtedy na pewno szkola nie ma klasy
 
         // when
-        assertThat(school.getSchoolClasses().size()).isEqualTo(0);
+        var result = school.removeClass(clazzA);
 
         // then
-        assertFalse(school.removeClass(clazzA));
+        assertFalse(result);
     }
     @Test
     public void shouldReturnTotalSchoolHeadcount(){
         // given
         school.addClass(clazzA);
 
         // when
+        // cos tutaj jest nie w porzadku - getHeadcount nie powinna miec argumentu
+        // tak byloby lepiej
+        // var resultAk = school.getHeadcount();
         boolean result = school.getHeadcount(school.getSchoolClasses()) == 1;
 
         // then
         assertTrue(result);
         assertThat(school.getSchoolClasses().size()).isEqualTo(1);
         // która z tych dwóch asercji powyżej jest lepsza do sprawdzenia żądanej metody?
+
+        // wg mnie taka :)
+        // assertThat(resultAk).isEqual(1);
     }
     @Test
-    public void shouldReturnMapOfClassesWithHeadcount(){
+    public void shouldReturnHeadcountPerClass(){
         // given
         school.addClass(clazzA);
 
         // when
-        school.getEachClassHeadcount(school.getSchoolClasses());
+        var result = school.getEachClassHeadcount(school.getSchoolClasses());
 
         //then
-     //   assertTrue(schoolClassesWithHeadcountMap.size() == 1);
+//        assertTrue(schoolClassesWithHeadcountMap.size() == 1);
+        assertThat(result).isNotNull()
```

Comment:
BTW nie do konca rozumialam co mialo byc w tym tescie. Oprocz tego co jest wychodzi mi, ze trzeba by jeszcze sprawdzic liczbe uczniow w poszczegolnych klasa.

Reply:
ten ostatni test nie działa jako że mam błąd w metodzie klasy School:
public Map<String, Integer> getEachClassHeadcount(Map<String, SchoolClass> schoolClasses)

i nie potrafię go zidentyfikować - dostaje Null Pointer Exception, możesz zerknąć na metodę w klasie?
a test co do zasady powiniem zwrócić mapę przedstawiającą parę nazwa klasy - liczba uczniów, ale mam null'a

Reply:
u mnie dziala z ta asercja jaka dalam ale jak pisalam wyzej niekoniecznie logika dziala :). Ta metoda i getHeadcount zdecydowanie wymaga wiekszej liczby testow. Wracajac to tej logiki - na razie implementacja jest bardzo prosta i nie jest przygotowa na te opisane wyzej przypadki specjalne. Co jezeli dodasz do szkoly klase bez uczniow - zakladam, ze wtedy wlasnie bedzie NPE. Mapa (nie kazda!) moze miec null jako value. Jezeli wybierzesz taka, ktora to dopuszcza takie akcje:
entry.getValue().getClassGroup().size();
trzeba zabezpieczyc przez nullem. NPE tutaj moze byc az w dwoch miejscach:
entry.getValue() - jak w szkole nie bedzie klasy
entry.getValue().getClassGroup() - jak w klasie nie bedzie uczniow

W logice, ktora dziala na bazie elementow zaleznych (tutaj w School sa SchoolClass, ktora z kolei ma List<Pupil>) trzeba sie przygotowac na rozne stany. 
W takich przypadkach robie takie cos:
- najpierw robie testy w SchoolClass - przy okazji robie testy ktore odopwiadaja na pytania: co bedzie jak nie bedzie uczniow, czy mam wszystkie metody potrzebne do operowaniu na klasie (moze warto dodac metode - hasPupils?). Prawdopodobnie w trakcie tych testow wyszloby mi, ze musze zrobic krok wstecz i zrobic testy na Pupil (i/lub Guardian) - czy nie mam logiki za bardzo skomplikowanej (czy na pewno pupil potrzebuje Guardian?)
- jak mam dobrze ja przetestowana (i ewentualnie jej zalezne klasy) to wtedy niektore elementy w school mi sie uproszcza (zakladam, ze na etapie testow SchoolClass i jej zaleznosci niektore zwiazki sie uproszcza, plus dodam nowe akcje na tych klasach - obie te zmiany uproszcza budowe school) i w efekcie testy jej beda prostsze. A jezeli nadal mam klopot z przygotowanie testu albo wykonaniem prostej walidacji w asercji to patrze co mam do poprawki w klasie. Oryginalne komentarze powstaly glowny w oparciu o ten ostatni punkt.